src/dhcp/clientv4.rs

```diff
@@ -0,0 +1,322 @@
+                p_state.retry = 0;
+                None
+            }
+            _ => None
+        }.map(|new_state| self.state = new_state);
+
+        config
+    }
+
+    fn egress<DeviceT: for<'d> Device<'d>>(&mut self, iface: &mut Interface<DeviceT>, udp_socket: &mut UdpSocket, now: Instant) -> Result<()> {
+        self.transaction_id += 1;
+        let mac = iface.ethernet_addr();
+        let mut dhcp_repr = DhcpRepr {
+            message_type: DhcpMessageType::Discover,
+            transaction_id: self.transaction_id,
+            client_hardware_address: mac,
+            client_ip: Ipv4Address::UNSPECIFIED,
+            your_ip: Ipv4Address::UNSPECIFIED,
+            server_ip: Ipv4Address::UNSPECIFIED,
+            router: None,
+            subnet_mask: None,
+            relay_agent_ip: Ipv4Address::UNSPECIFIED,
+            broadcast: false,
+            requested_ip: None,
+            client_identifier: Some(mac),
+            server_identifier: None,
+            parameter_request_list: None,
+            dns_servers: None,
+        };
+
+        match self.state {
+            ClientState::Discovering => {
+                self.next_egress = now + Duration::from_secs(DISCOVER_TIMEOUT);
+
+                let endpoint = IpEndpoint {
+                    addr: Ipv4Address::BROADCAST.into(),
+                    port: UDP_SERVER_PORT,
+                };
+                net_trace!("DHCP send discover to {}: {:?}", endpoint, dhcp_repr);
+                send_packet(udp_socket, &endpoint, &dhcp_repr)
+            }
+            ClientState::Requesting(ref mut r_state) if r_state.retry < REQUEST_RETRIES => {
+                r_state.retry += 1;
+                self.next_egress = now + Duration::from_secs(REQUEST_TIMEOUT);
+
+                let endpoint = IpEndpoint {
+                    addr: Ipv4Address::BROADCAST.into(),
+                    port: UDP_SERVER_PORT,
+                };
+                let requested_ip = match iface.ipv4_addr() {
+                    Some(addr) if !addr.is_unspecified() =>
+                        Some(addr),
+                    _ =>
+                        None,
+                };
+                dhcp_repr.message_type = DhcpMessageType::Request;
+                dhcp_repr.requested_ip = requested_ip;
+                dhcp_repr.server_identifier = Some(r_state.server_identifier);
+                dhcp_repr.parameter_request_list = Some(PARAMETER_REQUEST_LIST);
+                net_trace!("DHCP send request to {} = {:?}", endpoint, dhcp_repr);
+                send_packet(udp_socket, &endpoint, &dhcp_repr)
+            }
+            ClientState::Requesting(_) => {
+                net_debug!("DHCP request retries exceeded, restarting discovery");
+                // Timeout, restart discovery
+                self.state = ClientState::Discovering;
+                // Recurse to send discovery packet
+                self.egress(iface, udp_socket, now)
```

Comment:
This function has a rather large stack frame. Do you think you could replace recursion with a loop here?

Reply:
Oops, I wrongly assumed Rust had Tail-Call Optimization.

This function is now free of any recursions and loops.